### PR TITLE
Csv validate geos

### PR DESCRIPTION
**Pull request by jessihamel**
• Handles geos with 0, negative, or NaN by warning the user, then removing from data. In order to make the cartogram work, we remove it from the map geometry before calling the cartogram.
• Handles extra geos (fips for places we don't have or non-existent fips) by warning the user and removing from data. This fixes a bug where bad fips codes wouldn't throw an error, but would mess up the metric calculations.

Attached image shows cartogram with no Alabama or California.
screen shot 2016-10-12 at 3 12 55 pm

### CODE REVIEW

File: source/resources/DatasetResource.js [hunk 1]
```diff
@@ -24,7 +26,37 @@ class DatasetResource {
...
+    const valueIdString = badValueIds.join(', ')
+    let alertString = ''
+    if (mapIdString) {
+      alertString += `There is no associated map data associated with id(s): ${mapIdString}.`
+    }
+    if (valueIdString) {
+      alertString += ` Ids ${valueIdString} have zero or negative value.`
+    }
+    alertString += ' This data has been pruned.'
+    alert(alertString)
```
Comment (adamflorin) [inline source/resources/DatasetResource.js:60]: Let's put a `// eslint-disable-next-line no-alert` above this line to keep the linter happy

File: source/graphics/MapGraphic.js [hunk 2]
```diff
@@ -34,9 +34,23 @@ export default class MapGraphic extends Graphic {
...
+
+    // for custom uploads with incomplete data
+    if (properties.length !== baseMapTopoJson.objects.states.geometries.length) {
+      const statesWithData = properties.map(property => property[0])
+      filteredGeometries = baseMapTopoJson.objects.states.geometries
+        .filter(geom => statesWithData.indexOf(geom.id) > -1)
+      filteredTopoJson = JSON.parse(JSON.stringify(baseMapTopoJson)) // clones the baseMap
+      // only pass filtered geometries to topogram generator
+      filteredTopoJson.objects.states.geometries = filteredGeometries
+    }
```
Comment (adamflorin) [inline source/graphics/MapGraphic.js:64]: Can we factor this filtering logic out into a helper method to keep `computeCartogram` short and high-level?

File: source/components/HexMetrics.js [hunk 3]
```diff
@@ -102,11 +102,13 @@ export default class HexMetrics extends React.Component {
       const warn = (count.idealNHex === 0 && count.nHex === 0) ?
         <i className='fa fa-exclamation-triangle' /> :
         null
+      let className = count.deviation === 0 ? 'fade' : ''
+      if (count.disable) { className += ' disabled' }
```
Comment (adamflorin) [inline source/components/HexMetrics.js:107]: Can we get some linebreaks in that block? Should probably set up a linter rule for it…

File: source/resources/DatasetResource.js [hunk 4]
```diff
@@ -24,7 +26,37 @@ class DatasetResource {
...
-    return csvParseRows(csv, d => [d[0], +d[1]])
+    const features = mapResource.getUniqueFeatureIds()
+    const badMapIds = []
+    const badValueIds = []
+    const parsed = csvParseRows(csv, d => [d[0], +d[1]]).filter(row => {
+      const hasId = features.indexOf(row[0]) > -1
+      if (!hasId) {
+        badMapIds.push(row[0])
+      }
+      if (row[1] <= 0 || isNaN(+row[1])) {
```
Comment (adamflorin) [inline source/resources/DatasetResource.js]: Wasn't `row[1]` already parsed as an int in `csvParseRows` above?

Also, using `+` to do type casting still seems sneaky to me but I know how u D3 coders roll… ;)

**Review by adamflorin — APPROVED** [review]

File: source/resources/DatasetResource.js (hunk 4 above)
Comment (jessihamel) [reply source/resources/DatasetResource.js]: Good point, it's already parsed as int. Do you prefer ParseFloat to + for type casting? I didn't think there was much of a difference.

**Comment by adamflorin** [comment]
I don't think there's much functional difference, but that do like `parseFloat` for being explicit

**Pull request merged**